bazel/spec-bundling/spec-bundle.bzl

```diff
@@ -0,0 +1,106 @@
+load("@build_bazel_rules_nodejs//:index.bzl", "js_library")
+load("//bazel/esbuild:index.bzl", "esbuild", "esbuild_amd", "esbuild_config", "esbuild_esm_bundle")
+load("//bazel/spec-bundling:spec-entrypoint.bzl", "spec_entrypoint")
+load("//bazel/spec-bundling:bundle-config.bzl", "spec_bundle_config_file")
+
+def spec_bundle(
+        name,
+        deps,
+        platform,
+        bootstrap = [],
+        run_angular_linker = False,
+        linker_unknown_declaration_handling = None,
+        # We cannot use `ES2017` or higher as that would result in `async/await` not being downleveled.
+        # ZoneJS needs to be able to intercept these as otherwise change detection would not work properly.
+        target = "es2016",
+        external = [],
+        workspace_name = None):
+    """Macro that will bundle all test files with their respective transitive dependencies.
+
+    Bundled specs end up in a single bundle file that can be loaded within Karma or NodeJS directly.
+    This is helpful as Angular framework packages do not ship UMD files and to avoid overall
+    complexity with maintaining a runtime loader such as RequireJS or SystemJS.
+
+    Args:
+        name: Name of the spec bundle target
+        deps: Targets that contain all spec files. Files ending with `spec.js`
+          are picked up.
+        platform: Platform for which spec should be bundled. i.e. `node`, `browser` or `cjs-legacy`.
+        bootstrap: Targets providing bootstrap scripts that run before the specs. Files
+          ending with `init.js` are picked up.
+        target: Target ECMAScript to use for the specs bundle.
+        run_angular_linker: Whether the Angular linker should process the bundled code.
+        linker_unknown_declaration_handling: Control how unknown partial declarations should be
+          treated. This passes through to the `unknownDeclarationVersionHandling` linker plugin option.
+        external: List of modules/packages which should not be bundled.
+        workspace_name: Workspace name that needs to be provided for the AMD module name.
+    """
+
+    is_browser_test = platform == "browser"
+    is_legacy_cjs = platform == "cjs-legacy"
+    is_node = platform == "node"
+
+    package_name = native.package_name()
+    esbuild_attrs = dict()
+
+    spec_entrypoint(
```

Comment:
Note: This code is totally untouched but just has been moved from `index.bzl` into a new file.